### ISSUE TITLE
Heap snapshot is missing ability to navigate around

### ISSUE THREAD

**Issue by rmacnak-google** [flutter/devtools#4757]
Cannot look at an object and follow successors, predecessors, dominator parent, dominator children, etc.

**Comment by polina-c** [comment]
@rmacnak-google 

Will memory analysis tool meet your goals: 

```
% git clone https://dart.googlesource.com/sdk.git
% cd sdk/runtime/tools/heapsnapshot
sdk/runtime/tools/heapsnapshot % dart pub get
sdk/runtime/tools/heapsnapshot % dart bin/explore.dart
The Dart VM *.heapsnapshot analysis tool
Type `exit` or use Ctrl+D to exit.

(hsa) load <*.heapsnapshot>
```

See video demo at timestamp 3:28 of [go/day-1:-dash-team-summit-cam](https://goto.google.com/day-1:-dash-team-summit-cam)

**Comment by polina-c** [comment]
cc @mkustermann

**Comment by polina-c** [comment]
We could reference the tool in DevTools fro those who want to explore the snapshot.

**Comment by rmacnak-google** [comment]
No, the existence of the ability to navigate snapshots in tools/heapsnapshot, Graph Explorer and Observatory does not fix its absence in DevTools. The ability to [just look at what's in the snapshot](https://www.google.com/search?q=Richard+Feynman+just+look+at+the+thing) is the fundamental operation. Analysis is secondary.

**Comment by mkustermann** [comment]
There are close parallels between examining an application live and examining a snapshot of a live application state (which we use heap snapshots for).

In the ideal case I'd say that most/all debugger features we have (for a live application) could also be implemented for a snapshot of a live application state (a heapsnapshot). Inspecting an object is one such thing. The only? exception would be debugger features that require actually executing dart code.

While making the [runtime/tools/heapsnapshot](https://github.com/dart-lang/sdk/tree/main/runtime/tools/heapsnapshot) tool I have run into some limitations of the heap snapshot format that makes it difficult to build some of those features. I think it may be a good idea to extend the heapsnapshot format a little to support more such use cases (filed  https://github.com/dart-lang/sdk/issues/50546 for what came to my mind so far). Some things I had in mind were already done by @rmacnak-google - e.g. adding Smis in [cl/265562](https://dart-review.googlesource.com/c/sdk/+/265562)

**Comment by mkustermann** [comment]
> Cannot look at an object and follow successors, predecessors, dominator parent, dominator children, etc.

As a side note:

Personally I have found that dominator functionality of individual objects to be of limited use. If there's e.g. a string object that happens to be referred to by two objects, one may get a dominator that doesn't tell anything about how the string is referenced (dominator may be `ObjectPool` or something else very high up).
=> For individual objects I found shortest-retaining-path a much more useful functionality.

For aggregate views on memory the dominator-based graphexplorer functionality of boxes is more useful. Though it would be good to be able to make a heapsnapshot generation configurable to exclude VM-internal objects. That would prevent one to see `ObjectPool` or other such things dominating a lot of other objects (one of bullet points in https://github.com/dart-lang/sdk/issues/50546 I filed)

**Comment by polina-c** [comment]
We can think of reusing/adopting an existing graph exploration language/tool.

For example, https://neo4j.com/developer/cypher/ to query

**Comment by polina-c** [comment]
@bkonyi , can you itemize use cases and attach video demo on how observatory address them?

**Comment by polina-c** [comment]
Supported formats:
https://gephi.org/users/supported-graph-formats/
